### PR TITLE
stm32: clock_control: Enforce HCLK prescaler value

### LINKED ISSUE #17188: k_uptime_delta returns wrong times
I like to detect button events. They are selected after different holding times for a button.

A gpio interrupt is generated at every change of the button pin state. At the interrupt handler the time difference is measured between the current and the previous interrupt.
For that `k_uptime_delta` is used. I recognized that my measured times are at least by the factor 2 to low.

My setup is a own stm32F412 board with following config:
```
CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=96000000
CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000

# clock configuration
CONFIG_CLOCK_CONTROL=y

# Clock configuration for Cube Clock control driver
CONFIG_CLOCK_STM32_HSE_CLOCK=24000000
CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y

# use HSE as PLL input
CONFIG_CLOCK_STM32_PLL_SRC_HSE=y

# the board have an external oscillator
CONFIG_CLOCK_STM32_HSE_BYPASS=n

# produce 96MHz clock at PLL output
CONFIG_CLOCK_STM32_PLL_M_DIVISOR=12
CONFIG_CLOCK_STM32_PLL_N_MULTIPLIER=96
CONFIG_CLOCK_STM32_PLL_P_DIVISOR=2
CONFIG_CLOCK_STM32_PLL_Q_DIVISOR=4

CONFIG_CLOCK_STM32_AHB_PRESCALER=2
CONFIG_CLOCK_STM32_APB1_PRESCALER=4
CONFIG_CLOCK_STM32_APB2_PRESCALER=4
```
Maybe I setup the config wrongly or the `k_uptime_delta` function is buggy. Any ideas?

### PR DESCRIPTION
**Pull request by erwango**
STM32 clock control subsystem allows to configure a different
frequency value for core clock (SYSCLK) and AHB clock (HCLK).
Though, it is HCLK which is used to feed Cortex Systick timer
which  is used in zephyr as reference system clock.
If HCLK frequency is configured to a different value from SYSCLK
frequency, whole system is exposed to desynchro between zephyr clock
subsytem and STM32 HW configuration.
To prevent this, and until zephyr clock subsystem is changed to be
aware of this potential configuration, enforce AHB prescaler value
to 1 (which is current default value in use for all STM32 based
boards).

On STM32H7, enforce D1CPRE which fills the same role as ABH precaler.

On STM32MP1, the equivalent setting is done on A7 core, so it is
not exposed to the same issue as long as SYS_CLOCK_HW_CYCLES_PER_SEC
is set with the 'mlhclk_ck' clock frequency value. Update
matching boards documentation.

Fixes #17188

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

### CODE REVIEW

**Comment by erwango** [comment]
Adding the backport tag since this is valid on 1.14 branch as well.
Though, commit will have to be reworked since STM32H7 is not supported on 1.14

**Review by arnopo — APPROVED** [review]

File: boards/arm/96b_avenger96/doc/index.rst [hunk 1]
```diff
@@ -185,7 +185,8 @@ Default Zephyr Peripheral Mapping:
 System Clock
 ------------
 
-The Cortex®-M4 Core is configured to run at a 209 MHz clock speed.
+The Cortex®-M4 Core is configured to run at a 209 MHz clock speed. This value
+should match the configured mlhclk_ck frequency.
```
Comment (dbkinder) [inline boards/arm/96b_avenger96/doc/index.rst]: should this be ``must match`` (is it OK if it doesn't?)

File: boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst [hunk 2]
```diff
@@ -196,7 +196,8 @@ Default Zephyr Peripheral Mapping:
 System Clock
 ------------
 
-The Cortex®-M4 Core is configured to run at a 209 MHz clock speed.
+The Cortex®-M4 Core is configured to run at a 209 MHz clock speed. This value
+should match the configured mlhclk_ck frequency.
```
Comment (dbkinder) [inline boards/arm/stm32mp157c_dk2/doc/stm32mp157_dk2.rst]: Same here, should this be ``must match``?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/arm/96b_avenger96/doc/index.rst (hunk 1 above)
Comment (erwango) [reply boards/arm/96b_avenger96/doc/index.rst]: it must indeed, txs

**Comment by erwango** [comment]
@dbkinder, doc updated, txs for the review

**Review by dbkinder — APPROVED** [review]
+1 for doc changes, thanks.

**Pull request merged**